zebra-network/src/address_book.rs

```diff
@@ -219,43 +248,93 @@ impl AddressBook {
         self.by_addr.len()
     }
 
-    /// Update the metrics for this address book.
-    fn update_metrics(&self) {
-        let _guard = self.span.enter();
-
+    /// Returns metrics for the addresses in this address book.
+    pub fn address_metrics(&self) -> AddressMetrics {
         let responded = self.state_peers(PeerAddrState::Responded).count();
         let never_attempted = self.state_peers(PeerAddrState::NeverAttempted).count();
         let failed = self.state_peers(PeerAddrState::Failed).count();
-        let pending = self.state_peers(PeerAddrState::AttemptPending).count();
+        let attempt_pending = self.state_peers(PeerAddrState::AttemptPending).count();
 
         let recently_live = self.recently_live_peers().count();
         let recently_stopped_responding = responded
             .checked_sub(recently_live)
             .expect("all recently live peers must have responded");
 
+        AddressMetrics {
+            responded,
+            never_attempted,
+            failed,
+            attempt_pending,
+            recently_live,
+            recently_stopped_responding,
+        }
+    }
+
+    /// Update the metrics for this address book.
+    fn update_metrics(&mut self) {
+        let _guard = self.span.enter();
+
+        let m = self.address_metrics();
+
+        // TODO: rename to address_book.[state_name]
+        metrics::gauge!("candidate_set.responded", m.responded as f64);
+        metrics::gauge!("candidate_set.gossiped", m.never_attempted as f64);
+        metrics::gauge!("candidate_set.failed", m.failed as f64);
+        metrics::gauge!("candidate_set.pending", m.attempt_pending as f64);
```

Comment:
musing: I wonder if you could reasonably make a derive macro for recording all the fields of a struct as metrics, something like:

```rust
#[derive(Debug, Metrics)]
pub struct AddressMetrics {
    /// The number of addresses in the `Responded` state.
    #[metrics(gauge)]
    responded: usize,

    /// The number of addresses in the `NeverAttempted` state.
    #[metrics(gauge)]
    never_attempted: usize,

    /// The number of addresses in the `Failed` state.
    #[metrics(gauge)]
    failed: usize,

    /// The number of addresses in the `AttemptPending` state.
    #[metrics(gauge)]
    attempt_pending: usize,

    /// The number of `Responded` addresses within the liveness limit.
    #[metrics(gauge)]
    recently_live: usize,

    /// The number of `Responded` addresses outside the liveness limit.
    #[metrics(gauge)]
    recently_stopped_responding: usize,
}
```

then here you'd write something like:

```rust
m.report_metrics("candidate_set");
```

Definitely not worth implementing for this, but a fun idea none the less :eyes: